pkg/codegen/dotnet/gen_program.go

```diff
@@ -73,6 +73,15 @@ type generator struct {
 	// Program generation options
 	generateOptions GenerateProgramOptions
 	isComponent     bool
+	// when creating a list of items, we need to know the type of the list
+	// if is it a plain list, then `new()` should be used because we are creating List<T>
+	// however if we have InputList<T> or anything else, we use `new[]` because InputList<T> can be implicitly casted
+	// from an array
+	listInitializer string
```

Comment:
Bit grim this goes on the generator state? Can't we work this out, and thread it down more locally?

Reply:
> Bit grim this goes on the generator state?

Yes

> Can't we work this out, and thread it down more locally

Not unless we rewrite `ExpressionGenerator` because it not possible to know at the point of generating an expression (i.e. `GenTupleConsExpxression` for arrays) what context information about that expression is available 😞